### ISSUE TITLE
mat: allow to explicitly specify SVDKind for both jobU, jobVT in Factorize()

### ISSUE THREAD

**Issue by aleh-null** [gonum/gonum#487]
I'm implementing Kalman Filter algorithm based on singular value decomposition. The crucial part of this algorithm involves the computation of the singular value of an asymmetric matrix through SVD without explicitly forming its left factor (U). In other words, the algorithm only uses singular values and right singular vectors (Vt). At the moment, I cannot find a way to only get the parts described. I think It would be a significant improvement to `Factorize()` if it used separate arguments for each part, as it is done f.ex in R's `svd()`.

I also tried to write a my own `Factorize()` that wraps `lapack64.Gesvd()`, however private `getFloats()` and `putFloats()` keep me from doing this. Maybe there is an option to make those public?

**Comment by kortschak** [comment]
Do you have a suggestion for the API? Adding multiple parameters to the `Factorize` call has always struck me as pretty ugly.

`getFloats` and `putFloats` are unlikely to be made public. We have considered placing them in more widely available internal packages though that leads to some problems. The argument against making them publicly available is that it is easy to retain a value that is put and cause unpleasant and difficult to find aliasing problems.

**Comment by btracey** [comment]
@aleh-null : In the short term, putFloats and getFloats shouldn't stop you from your own wrapper, you can swap them out for `make`. 

The lapack call has 4 options for each of the vectors, but one of them is just a memory optimization. So really there are 9 options. We currently support 3 of those 9 (where the two SVDKinds are equal). It seems like the easiest way is to just add the remaining options.
```
SVDNoneThin
SVDNoneFull
SVDThinNone
SVDThinFull
SVDFullNone
SVDFullThin
```
I wonder if all of these are really necessary though. Having Thin or None is just an optimization. For this reason, I can't see why one would be justified in `ThinFull` or `FullThin`. Similarly, it seems that if you don't want to spend the time/memory to compute one of the factors, you wouldn't want to spend the memory to compute the full factorization with a bunch of zeros. So I wonder if we can get away with just adding `SVDNoneThin` and `SVDThinNone`, which is not too big of an extension to the API surface.

**Comment by btracey** [comment]
(If we're just adding those two, and only those two, we could rename ThinU and ThinV)

**Comment by aleh-null** [comment]
When it comes to `ThinFull` or `FullThin` I also don't find any use case. Both of them are meaningless for square matrixes, also can be confusing for rectangular ones. If M > N `ThinFull` is the same as `ThinThin` and if M < N then `ThinFull` is the same as `FullFull` etc. When in comes to `NoneThin` and `NoneFull` the first one will definitely work for me and I could not find any usage of the later one. I also think that `NoneThin/Full` options will most likely be used for nonnegative definite matrixes where Full/Thin is the same.

**Comment by btracey** [comment]
Sorry, what do you mean by "stick to the lapack naming"?

**Comment by aleh-null** [comment]
As you said:
> So I wonder if we can get away with just adding `SVDNoneThin` and `SVDThinNone`, which is not too big of an extension to the API surface.
I think that those 2 should be sufficient. You also proposed to name those as `ThinU` and `ThinV`, however I prefer the former.

**Comment by kortschak** [comment]
I'd like to suggest a bit set.

```
const (
    thin = 1 << iota // (0 -> full)
    u                // (0 -> no U)
    v                // (0 -> no V)
)
```

This maps as follows: `0` or `thin` -> `SVDNone`, `u|v` -> `SVDFull`, `u|thin` -> `SVDThinNone`, `v|thin` -> `SVDNoneThin`, `u|v|thin` -> `SVDThin`, and `u` and `v` -> currently not implemented.

**Comment by kortschak** [comment]
I disagree; the use of bitfields for options is widespread (not [just](https://golang.org/pkg/log/#pkg-constants) [in](https://golang.org/pkg/text/tabwriter/#pkg-constants) [the](https://golang.org/pkg/regexp/syntax/#Flags) [unix](https://golang.org/pkg/os/#OpenFile) [world](https://golang.org/pkg/os/#FileMode), [but](https://golang.org/pkg/go/ast/#MergeMode) [in](https://golang.org/pkg/go/printer/#Mode) [the](https://golang.org/pkg/go/doc/#Mode) [stdlib](https://golang.org/pkg/go/parser/#Mode)).

If I see
```
var svd mat.SVD
ok := svd.Factorize(a, mat.SVDThinNone)
```
and 
```
var svd mat.SVD
ok := svd.Factorize(a, mat.SVDNoneThin)
```
I need to go to the documentation for `SVDKind` to find out which matrix is not being retained because it is positional.

If I see
```
var svd mat.SVD
ok := svd.Factorize(a, mat.SVDU|mat.SVDThin)
```
and 
```
var svd mat.SVD
ok := svd.Factorize(a, mat.SVDV|mat.SVDThin)
```
I can immediately see that the first is getting U and it is thin, the seond is getting V and it is thin (In writing the example above with `SVDNoneThin` I had intended to write the same semantics as the bitfields and got it wrong - not trying to).

It also requires fewer constants for nearly all of the cases (and arguably the cases that are needed/wanted). With a (I think unnecessary) increase in complexity it gets all cases with 6 necessary constants (and one convenience constant).
